File: golang/799.champagne-tower.go
func champagneTower(poured int, query_row int, query_glass int) float64 {
    // dp[i][j] - designates how much liquid passed through
    // j-th glass in i-th row
    dp := make([][]float64, query_row + 1)
    dp[0] = make([]float64, 1)
    dp[0][0] = float64(poured)

    var left, right, left_extra, right_extra float64

    for row := 1; row <= query_row; row++ {
        dp[row] = make([]float64, row + 1)
        for col := 0; col <= row; col++ {
            // get how much liquid passed through left ancestor
            if col > 0 {
                left = dp[row - 1][col - 1]
            } else {
                left = 0.0
            }
            // get how much liquid passed through right ancestor
            if col < row {
                right = dp[row - 1][col]
            } else {
                right = 0.0
            }
            // calculate how much extra liquid passed through left and right ancestor
            left_extra = max(0.0, left - 1) // take 1 to fill glass to the edge
            right_extra = max(0.0, right - 1) // take 1 to fill glass to the edge

            // half of extra liquid will be spilled to current glass
            // and othe half to it's neighbours
            dp[row][col] = left_extra / 2 + right_extra / 2
        }
    }

    // convert how much liquid passed through the glass
    // to how much liquid left in a glass
    return min(1.0, dp[query_row][query_glass])
}

// min float64
func min(a, b float64) float64 {
    if a < b {
        return a
    }
    return b
}

// max float64
func max(a, b float64) float64 {
    if a > b {
        return a
    }
    return b
}

File: python/799.champagne-tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:

        dp = [[0] * (query_row + 1) for _ in range(query_row + 1)]
        dp[0][0] = poured
        
        for row in range(1, query_row + 1):
            for i in range(row + 1):
                # how much liquid went through left and right ancestor
                left = dp[row - 1][i - 1] if i > 0 else 0
                right = dp[row - 1][i] if i <= row else 0
                # how much extra liquid spilled over 
                left_extra = max(0, left - 1)
                right_extra = max(0, right - 1)
                # take half of left ancestor and half of right ancestor
                # half is because ancestor have two edges to spill extra liquid
                dp[row][i] = left_extra / 2 + right_extra / 2

        return min(1.0, dp[query_row][query_glass])